### PR TITLE
Add impl stage-testing,image-consistency-check,image-signed-check cmds

### PR DESCRIPTION
**Pull request by wewang58**
Add call jenkins stage pipe job func. 
next will confirm it added to which part, after all advisorifes push to cdn stage success to call it  or just wait metadata advisory push to cdn stage finish, then call it?

### CODE REVIEW

File: oar/core/config_store.json [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
     "build_data_url": "https://raw.githubusercontent.com/openshift-eng/ocp-build-data/%s/releases.yml",
     "jira_server": "https://issues.redhat.com",
+    "jenkins_server": "https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/zstreams/job/Stage-Pipeline/buildWithParameters",
```
Comment (rioliu-rh) [inline oar/core/config_store.json:4]: @wewang58 suggest to save jenkins server base url here

File: oar/core/jenkins_trigger.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import requests
+import logging
+from oar.core.config_store import ConfigStore
+from oar.core.exceptions import JenkinsTriggerException
+
+logger = logging.getLogger(__name__)
+class JenkinsTrigger:
+    def __init__(self, cs: ConfigStore):
+        self._cs = cs
+        self.version = re.match("(\d).(\d+)", self._cs.release)
```
Comment (rioliu-rh) [inline oar/core/jenkins_trigger.py:11]: @wewang58 you can use `util.get_y_release`

**Comment by wewang58** [comment]
@rioliu-rh Updated according comments, please review it again, thanks

File: oar/cli/cmd_stage_testing.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    """
+    # get config store from context
+    cs = ctx.obj["cs"]
+    try: 
+        report = WorksheetManager(cs).get_test_report()
+        cdn_result = report.get_task_status(LABEL_TASK_PUSH_TO_CDN)
+        if cdn_result == "Pass":
+            jh = JenkinsHelper(cs)
+            build_url = jh.call_stage_job(jh.url, jh.username,jh.token,jh.metadata_ad, jh.pull_spec)
+            logger.info(f"trigger stage pipeline build is: <{build_url}>")
```
Comment (rioliu-rh) [inline oar/cli/cmd_stage_testing.py:23]:
```suggestion
            logger.info(f"triggered stage pipeline job: <{build_url}>") 
```

File: oar/cli/cmd_stage_testing.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    cs = ctx.obj["cs"]
+    try: 
+        report = WorksheetManager(cs).get_test_report()
+        cdn_result = report.get_task_status(LABEL_TASK_PUSH_TO_CDN)
+        if cdn_result == "Pass":
+            jh = JenkinsHelper(cs)
+            build_url = jh.call_stage_job(jh.url, jh.username,jh.token,jh.metadata_ad, jh.pull_spec)
+            logger.info(f"trigger stage pipeline build is: <{build_url}>") 
+    except Exception as e:
+        logger.exception("pushing to cdn stages are not complete, cannot do stage testing")
```
Comment (rioliu-rh) [inline oar/cli/cmd_stage_testing.py:25]: oops, copy/paste line. let's adjust it accordingly

Comment (rioliu-rh) [inline oar/cli/cmd_stage_testing.py:25]: I did see any exception raising when `cdn_result` is not `Pass`.

File: oar/core/config_store.py [hunk 5]
```diff
@@ -195,6 +195,18 @@ def get_jira_server(self):
...
+        """
+        Get jenkins server url
+        """
+        return self._local_conf["jenkins_server"]
+
+    def get_jenkins_username(self):
+        """
+        Get jenkins username
+        """
+        return self._local_conf["jenkins_user"]
```
Comment (rioliu-rh) [inline oar/core/config_store.py:208]: jenkins user/token should be consistent, suggest to declare user as env var as well.

File: oar/core/jenkins_helper.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        self.pull_spec = "quay.io/openshift-release-dev/ocp-release:" +self._cs.release+ "-x86_64"
+        self.url = self._cs.get_jenkins_server()+ "/job/zstreams/"
+        self.username=self._cs.get_jenkins_username()
+        self.token=self._cs.get_jenkins_token()
+    def call_stage_job(url, username, token, version, metadata_ad, pull_spec): 
+        try:
+            server = jenkins.Jenkins(url, username=username, password=token)
+            server.build_job("Stage-Pipeline",parameters={'VERSION': version,'METADATA_AD':metadata_ad,'PULL_SPEC':pull_spec})
+        except jenkins.JenkinsException as ej:   
+            raise JenkinsHelperException("call stage pipeline job failed")
```
Comment (rioliu-rh) [inline oar/core/jenkins_helper.py:22]:
```suggestion
            raise JenkinsHelperException("call stage pipeline job failed") from ej
```

File: oar/core/jenkins_helper.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        last_build_number = server.get_job_info("Stage-Pipeline")['lastBuild']['number']
+        if not last_build_number:
+            logger.exception("Trigger stage pipeline job failed")
```
Comment (rioliu-rh) [inline oar/core/jenkins_helper.py:25]:
```suggestion
            raise JenkinsHelperException("Trigger stage pipeline job failed: cannot get latest build number")
```

File: tests/test_call_stage_job.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+class TestJenkinsHelper(unittest.TestCase):
+    def setUp(self):
+        self.cs = ConfigStore("4.12.11")
+        self.jh = JenkinsHelper(self.cs)
+
+    def test_init(self):
+        pass
+
+    def test_call_stage_job(self):
+        res = self.jt.call_stage_job(self.jh.url,"wewang", self.cs.get_jenkins_token(),self.jh.version,self.jh.metadata_ad, self.jh.pull_spec)
```
Comment (rioliu-rh) [inline tests/test_call_stage_job.py:15]: check the warning

File: oar/cli/cmd_stage_testing.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import logging
+from oar.core.jenkins_helper import JenkinsHelper
+from oar.core.const import *
+from oar.core.worksheet_mgr import WorksheetManager
+
+logger = logging.getLogger(__name__)
+
+@click.command()
+@click.pass_context
+def stage_testing(ctx):
```
Comment (rioliu-rh) [inline oar/cli/cmd_stage_testing.py:11]: still need to add this cmd to `cmd_group.py`

Comment (wewang58) [reply oar/cli/cmd_stage_testing.py:12]: Updated: Commands:
  change-advisory-status     Change advisory status e.g.
  check-greenwave-cvp-tests  Check Greenwave CVP test results for all...
  create-test-report         Create test report for z-stream release
  drop-bugs                  Drop bugs from advisories
  push-to-cdn-staging        Trigger push job for cdn stage targets
  stage-testing              Trigger stage pipeline test
  take-ownership             Take ownership for advisory and jira subtasks
  update-bug-list            Update bug status listed in report, update...

File: oar/core/config_store.py (hunk 5 above)
Comment (wewang58) [reply oar/core/config_store.py]: updated

File: oar/cli/cmd_stage_testing.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+            try: 
+                jh = JenkinsHelper(cs)
+                build_url = jh.call_stage_job(jh.url, jh.username,jh.token,jh.metadata_ad, jh.pull_spec)
+            except JenkinsHelperException as jh:
+                logger.exception("trigger stage pipeline job failed")
+                raise
+            logger.info(f"triggerred stage pipeline job: <{build_url}>") 
+            report.update_task_status(LABEL_TASK_STAGE_TEST, TASK_STATUS_INPROGRESS)
+        else:
+            report.update_task_status(LABEL_TASK_PUSH_TO_CDN, TASK_STATUS_FAIL)
```
Comment (rioliu-rh) [inline oar/cli/cmd_stage_testing.py:31]: i'm confused, why we need to mark job `push_to_cdn` failed here.

**Comment by wewang58** [comment]
@rioliu-rh tested in local, it can trigger stage job and update Stage testing to in progress
[wewang@fedora release-tests]$ python cli  -r 4.13.2 stage-testing
2023-06-08T10:56:30Z: ERROR: handle_other(): Mutual authentication failed
2023-06-08T10:56:33Z: INFO: triggerred stage pipeline job: <https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/zstreams/job/Stage-Pipeline/702>
2023-06-08T10:56:34Z: INFO: task [Stage testing] status is changed to [In Progress]
[wewang@fedora release-tests]$ python cli  -r 4.13.2 stage-testing
2023-06-08T11:03:32Z: ERROR: handle_other(): Mutual authentication failed
2023-06-08T11:03:36Z: INFO: triggerred stage pipeline job: <https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/zstreams/job/Stage-Pipeline/703>
2023-06-08T11:03:37Z: INFO: task [Stage testing] status is changed to [In Progress]

about "ERROR: handle_other(): Mutual authentication failed", it's python jenkins related bug, the bug is still new: https://bugs.launchpad.net/python-jenkins/+bug/1849302

File: oar/cli/cmd_stage_testing.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+def stage_testing(ctx):
+    """
+    Trigger stage pipeline test
+    """
+    # get config store from context
+    cs = ctx.obj["cs"] 
+    try:
+        report = WorksheetManager(cs).get_test_report()
+        cdn_result = report.get_task_status(LABEL_TASK_PUSH_TO_CDN)
+        if cdn_result == "Pass":
```
Comment (rioliu-rh) [inline oar/cli/cmd_stage_testing.py:21]: here, `Pass` can replaced by constant `TASK_STATUS_PASS`

File: oar/cli/cmd_stage_testing.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+            try: 
+                jh = JenkinsHelper(cs)
+                build_url = jh.call_stage_job(jh.url, jh.username,jh.token, jh.version, jh.metadata_ad, jh.pull_spec)    
+            except JenkinsHelperException as jh:
+                logger.exception("trigger stage pipeline job failed")
+                raise
+            logger.info(f"triggerred stage pipeline job: <{build_url}>") 
+            report.update_task_status(LABEL_TASK_STAGE_TEST, TASK_STATUS_INPROGRESS)
+        else:
+            logger.info("push to cdn stage is not complete, so will not trigger stage test")
```
Comment (rioliu-rh) [inline oar/cli/cmd_stage_testing.py:31]:
```suggestion
            logger.info("push to cdn stage is not completed, so will not trigger stage test")
```

**Comment by rioliu-rh** [comment]
> @rioliu-rh tested in local, it can trigger stage job and update "Stage testing" in report to "in progress"
> 
> [wewang@fedora release-tests]$ python cli -r 4.13.2 stage-testing 2023-06-08T11:03:32Z: ERROR: handle_other(): Mutual authentication failed 2023-06-08T11:03:36Z: INFO: triggerred stage pipeline job: [mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/zstreams/job/Stage-Pipeline/703](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/zstreams/job/Stage-Pipeline/703) 2023-06-08T11:03:37Z: INFO: task [Stage testing] status is changed to [In Progress]
> 
> about "ERROR: handle_other(): Mutual authentication failed", it's python jenkins related bug, the bug is still new: [bugs.launchpad.net/python-jenkins/+bug/1849302](https://bugs.launchpad.net/python-jenkins/+bug/1849302), it just give the error info, not affect connect to jenkins.

Can we send out the job url via slack

**Comment by rioliu-rh** [comment]
> > about "ERROR: handle_other(): Mutual authentication failed", it's python jenkins related bug, the bug is still new: [bugs.launchpad.net/python-jenkins/+bug/1849302](https://bugs.launchpad.net/python-jenkins/+bug/1849302), it just give the error info, not affect connect to jenkins.

you can disable logger of jenkins lib

**Comment by wewang58** [comment]
> > > about "ERROR: handle_other(): Mutual authentication failed", it's python jenkins related bug, the bug is still new: [bugs.launchpad.net/python-jenkins/+bug/1849302](https://bugs.launchpad.net/python-jenkins/+bug/1849302), it just give the error info, not affect connect to jenkins.
> 
> you can disable logger of jenkins lib

**Comment by wewang58** [comment]
> > > about "ERROR: handle_other(): Mutual authentication failed", it's python jenkins related bug, the bug is still new: [bugs.launchpad.net/python-jenkins/+bug/1849302](https://bugs.launchpad.net/python-jenkins/+bug/1849302), it just give the error info, not affect connect to jenkins.
> 
> you can disable logger of jenkins lib

sorry, you mean disable logger of jenkins_helper module, right? i did that, but still have the error info

**Comment by wewang58** [comment]
Add new cmds:image-consistency-check and image-signed-check cmds and get job status via job id:
[wewang@fedora release-tests]$ python cli -r 4.13.2 image-consistency-check
2023-06-13T09:18:13Z: INFO: job id is not set, will trigger image consistency check job
2023-06-13T09:18:17Z: ERROR: handle_other(): Mutual authentication failed
2023-06-13T09:18:20Z: INFO: triggerred image consistency check job: <https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/image-consistency-check/2229>
2023-06-13T09:18:21Z: INFO: task [Payload-Images Consistency Testing] status is changed to [In Progress]

[wewang@fedora release-tests]$ python cli -r 4.13.2 image-signed-check
2023-06-13T09:19:53Z: ERROR: handle_other(): Mutual authentication failed
2023-06-13T09:19:56Z: INFO: triggerred signature_check job: https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/zstreams/job/signature_check/137
2023-06-13T09:19:57Z: INFO: sent slack message to <#test-wewang>
2023-06-13T09:19:58Z: INFO: task [Payload image is well signed] status is changed to [In Progress]

[wewang@fedora release-tests]$ python cli -r 4.13.2 image-consistency-check -n 2229
2023-06-13T09:47:13Z: INFO: check image consistency check job status according  to job id:2229
2023-06-13T09:47:14Z: ERROR: handle_other(): Mutual authentication failed
2023-06-13T09:47:17Z: INFO: image-consistency-check job: 2229 status is: ABORTED

about jenkins error info, still need to time to check, will submit a new pr to fix it.

File: oar/cli/cmd_image_consistency_check.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        report = WorksheetManager(cs).get_test_report()
+        image_consistency_result = report.get_task_status(LABEL_TASK_IMAGE_CONSISTENCY_TEST)
+        if image_consistency_result == TASK_STATUS_PASS:
+            logger.info("image consistency check already pass, not need to trigger again")
+        elif image_consistency_result == TASK_STATUS_INPROGRESS:   
+            logger.info("job[image-consistency-check] already triggered and in progress, not need to trigger again")
+        else:
+            nm = NotificationManager(cs)
+            try: 
+                build_url = jh.call_image_consistency_job(jh.username,jh.token, jh.vversion, jh.errata_numbers, jh.pull_spec)
```
Comment (rioliu-rh) [inline oar/cli/cmd_image_consistency_check.py:33]: @wewang58 little weird, if the param values are internal fields of jenkins helper. why you call them explicitly here

Comment (wewang58) [reply oar/cli/cmd_image_consistency_check.py:33]: @rioliu-rh  let me check and update, thanks

Comment (wewang58) [reply oar/cli/cmd_image_consistency_check.py]: @rioliu-rh  Updated and tested it

**Pull request merged**